core/v2/serialization.go

```diff
@@ -0,0 +1,265 @@
+							Type: "uint256",
+						},
+						{
+							Name: "Y",
+							Type: "uint256",
+						},
+					},
+				},
+				{
+					Name: "lengthCommitment",
+					Type: "tuple",
+					Components: []abi.ArgumentMarshaling{
+						{
+							Name: "X",
+							Type: "uint256[2]",
+						},
+						{
+							Name: "Y",
+							Type: "uint256[2]",
+						},
+					},
+				},
+				{
+					Name: "lengthProof",
+					Type: "tuple",
+					Components: []abi.ArgumentMarshaling{
+						{
+							Name: "X",
+							Type: "uint256[2]",
+						},
+						{
+							Name: "Y",
+							Type: "uint256[2]",
+						},
+					},
+				},
+				{
+					Name: "length",
+					Type: "uint32",
+				},
+			},
+		},
+		{
+			Name: "quorumNumbers",
+			Type: "bytes",
+		},
+		{
+			Name: "paymentMetadataHash",
+			Type: "bytes32",
+		},
+	}
+}
+
+func (b *BlobHeader) toABIStruct() (abiBlobHeader, error) {
+	paymentHash, err := b.PaymentMetadata.Hash()
+	if err != nil {
+		return abiBlobHeader{}, err
+	}
+	return abiBlobHeader{
+		BlobVersion: uint8(b.BlobVersion),
+		BlobCommitments: abiBlobCommitments{
+			Commitment: abiG1Commit{
+				X: b.BlobCommitments.Commitment.X.BigInt(new(big.Int)),
+				Y: b.BlobCommitments.Commitment.Y.BigInt(new(big.Int)),
+			},
+			LengthCommitment: abiG2Commit{
+				X: [2]*big.Int{
+					b.BlobCommitments.LengthCommitment.X.A0.BigInt(new(big.Int)),
+					b.BlobCommitments.LengthCommitment.X.A1.BigInt(new(big.Int)),
+				},
+				Y: [2]*big.Int{
+					b.BlobCommitments.LengthCommitment.Y.A0.BigInt(new(big.Int)),
+					b.BlobCommitments.LengthCommitment.Y.A1.BigInt(new(big.Int)),
+				},
+			},
+			LengthProof: abiG2Commit{
+				X: [2]*big.Int{
+					b.BlobCommitments.LengthProof.X.A0.BigInt(new(big.Int)),
+					b.BlobCommitments.LengthProof.X.A1.BigInt(new(big.Int)),
+				},
+				Y: [2]*big.Int{
+					b.BlobCommitments.LengthProof.Y.A0.BigInt(new(big.Int)),
+					b.BlobCommitments.LengthProof.Y.A1.BigInt(new(big.Int)),
+				},
+			},
+			Length: uint32(b.BlobCommitments.Length),
+		},
+		QuorumNumbers:       b.QuorumNumbers,
+		PaymentMetadataHash: paymentHash,
+	}, nil
+}
+
+func (b *BlobHeader) BlobKey() (BlobKey, error) {
+	blobHeaderType, err := abi.NewType("tuple", "", blobHeaderArgMarshaling())
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	arguments := abi.Arguments{
+		{
+			Type: blobHeaderType,
+		},
+	}
+
+	s, err := b.toABIStruct()
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	bytes, err := arguments.Pack(s)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	var headerHash [32]byte
+	hasher := sha3.NewLegacyKeccak256()
+	hasher.Write(bytes)
+	copy(headerHash[:], hasher.Sum(nil)[:32])
+
+	return headerHash, nil
+}
+
+func (c *BlobCertificate) Hash() ([32]byte, error) {
+	if c.BlobHeader == nil {
+		return [32]byte{}, fmt.Errorf("blob header is nil")
+	}
+
+	blobCertType, err := abi.NewType("tuple", "", []abi.ArgumentMarshaling{
+		{
+			Name:       "blobHeader",
+			Type:       "tuple",
+			Components: blobHeaderArgMarshaling(),
```

Comment:
Wouldn't it make sense to just use the `BlobKey` here?

Reply:
As long as we stay consistent in contracts, we can do that. Then the hash will be applied to a struct like this instead of actual blob cert: 
```
struct {
  bytes32 blobHeaderHash;
  uint16[] relayKeys;
}
```

Reply:
It's probably cheaper to do it the way you have it here. Let's leave it for now. I can't think of any situations where this would matter.